### PR TITLE
Add LED 7-Segment HT16K33 driver.

### PR DESCRIPTION
**Pull request by simonlbn**
This driver allows controlling Adafruit 0.56" 4-Digit 7-Segment
Display w/I2C Backpack.

I tried to follow coding conversion from other files, but not sure how much I got it right.

Do comment on the patch etc. and I can fix things instead of you having to do it later.

For now only added it to the nodemcu board, as that's the one I have tested it on. That said, I see no reason it shouldn't work on other boards using i2c_soft driver.

PS. I did the work entirely in my own spare time, but the simple way to release patches like this means my dayjob retains copyright, and for code which is under MIT this shouldn't really matter.

### CODE REVIEW

File: doc/library-reference/drivers/led_7seg_ht16k33.rst [hunk 1]
```diff
@@ -0,0 +1,20 @@
+:mod:`led_7seg_ht16k33` --- LED 7-Segment HT16K33
+===========================
```
Comment (eerimoq) [inline doc/library-reference/drivers/led_7seg_ht16k33.rst]: Add more "=" to match title length.

File: doc/library-reference/drivers/led_7seg_ht16k33.rst [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+.. module:: led_7seg_ht16k33
+   :synopsis: LED Stuff.
```
Comment (eerimoq) [inline doc/library-reference/drivers/led_7seg_ht16k33.rst]: Change synopsis to "LED 7-Segment HT16K33.". Same test as in the title.

File: examples/led_7seg_ht16k33/config.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ * This file is part of the Simba project.
+ */
+
+#ifndef __CONFIG_H__
+#define __CONFIG_H__
+
+#define CONFIG_MINIMAL_SYSTEM                               1
+#define CONFIG_I2C_SOFT                                     1
+#define CONFIG_LED_7SEG_HT16K33                             1
+#endif
```
Comment (eerimoq) [inline examples/led_7seg_ht16k33/config.h]: Add a newline before #endif.

File: examples/led_7seg_ht16k33/main.c [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+struct led_7seg_ht16k33_driver_t led;
+struct i2c_soft_driver_t i2c;
+
+void counting(void)
+{
+    int i, mult, dot;
+
+    for (mult = 1, dot = 3; mult <= 0x1000; mult *= 0x10, dot--) {
+        led_7seg_ht16k33_show_dot(&led, dot, true);
+	for (i = 0; i <= 0xf; i++) {
```
Comment (eerimoq) [inline examples/led_7seg_ht16k33/main.c]: Add newline before and after for-statement.

File: examples/led_7seg_ht16k33/main.c [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+    led_7seg_ht16k33_show_colon(&led, true);
+    led_7seg_ht16k33_display(&led);
+    thrd_sleep(1);
+
+    led_7seg_ht16k33_show_colon(&led, false);
+    led_7seg_ht16k33_display(&led);
+    thrd_sleep(1);
+
+    for (int i = 0; i < 4; i++) {
```
Comment (eerimoq) [inline examples/led_7seg_ht16k33/main.c]: Declare "int i" at beginning of the function.

File: examples/led_7seg_ht16k33/main.c [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+    thrd_sleep(1);
+
+    led_7seg_ht16k33_set_num(&led, 0xfefe, 16);
+    led_7seg_ht16k33_display(&led);
+    thrd_sleep(1);
+}
+
+int main()
+{
+
```
Comment (eerimoq) [inline examples/led_7seg_ht16k33/main.c]: Remove newline.

File: examples/led_7seg_ht16k33/main.c [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    /* Initialization. */
+    sys_start();
+    led_7seg_ht16k33_module_init();
+    i2c_soft_init(&i2c, &pin_d5_dev, &pin_d6_dev, 50000, 1000000, 1000);
+    led_7seg_ht16k33_init(&led, &i2c, LED_7SEG_HT16K33_DEFAULT_I2C_ADDR);
+    led_7seg_ht16k33_start(&led);
+
+    std_printf("\n\n");
```
Comment (eerimoq) [inline examples/led_7seg_ht16k33/main.c]: Use "\r\n" as line sepatrator.

Also, wrap strings given to std_printf() in the OSTR() macro. All strings that are intended for humans should be wrapped in this macro. The macro is used by the build system to remove format strings from the binary and store them in an external database. For applications without SOAM, the macro expands to FSTR(), which is short for Far STRing. The std_printf() function takes a far string as format string. Far strings are stored in flash on some boards, and the application will crash trying to access far strings in RAM.

File: examples/led_7seg_ht16k33/main.c [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    std_printf(sys_get_info());
+    while (1) {
+	counting();
```
Comment (eerimoq) [inline examples/led_7seg_ht16k33/main.c]: Use four space indentation, not tab.

File: examples/led_7seg_ht16k33/main.c [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+    /* Initialization. */
+    sys_start();
+    led_7seg_ht16k33_module_init();
+    i2c_soft_init(&i2c, &pin_d5_dev, &pin_d6_dev, 50000, 1000000, 1000);
+    led_7seg_ht16k33_init(&led, &i2c, LED_7SEG_HT16K33_DEFAULT_I2C_ADDR);
+    led_7seg_ht16k33_start(&led);
+
+    std_printf("\n\n");
+    std_printf(sys_get_info());
+    while (1) {
```
Comment (eerimoq) [inline examples/led_7seg_ht16k33/main.c]: Add a newline before while-statement.

File: src/drivers/led_7seg_ht16k33.c [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+
+int led_7seg_ht16k33_set_num(struct led_7seg_ht16k33_driver_t *self_p,
+                             int num, int base)
+{
+    ASSERTN(self_p != NULL, EINVAL);
+    ASSERTN(base > 0, EINVAL);
+    ASSERTN(base <= 16, EINVAL);
+
+    for (int i = 0, j = 4; i < 4; i++, j--) {
+        if (i == SEVEN_SEG_COLON_POS)
```
Comment (eerimoq) [inline src/drivers/led_7seg_ht16k33.c]: Always use "{}", even for one line statements.

File: src/drivers/led_7seg_ht16k33.h [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+ * Number cannot be more than 4 digits AKA base^4 - 1.
+ *
+ * @param[in] self_p Driver object.
+ * @param[in] num Number to set.
+ * @param[in] base Base of \a num.
+ *
+ * @return zero(0) or negative error code.
+ */
+int led_7seg_ht16k33_set_num(struct led_7seg_ht16k33_driver_t *self_p,
+                             int num, int base);
```
Comment (eerimoq) [inline src/drivers/led_7seg_ht16k33.h]: One argument per line.

File: doc/library-reference/drivers/led_7seg_ht16k33.rst (hunk 1 above)
Comment (simonlbn) [inline doc/library-reference/drivers/led_7seg_ht16k33.rst]: Done.

File: doc/library-reference/drivers/led_7seg_ht16k33.rst (hunk 2 above)
Comment (simonlbn) [inline doc/library-reference/drivers/led_7seg_ht16k33.rst]: Doh, done.

File: examples/led_7seg_ht16k33/config.h (hunk 3 above)
Comment (simonlbn) [inline examples/led_7seg_ht16k33/config.h]: Done.

File: examples/led_7seg_ht16k33/main.c (hunk 4 above)
Comment (simonlbn) [inline examples/led_7seg_ht16k33/main.c]: Done (I think - if I understood you correctly).

File: examples/led_7seg_ht16k33/main.c (hunk 5 above)
Comment (simonlbn) [inline examples/led_7seg_ht16k33/main.c]: Done.

File: examples/led_7seg_ht16k33/main.c (hunk 7 above)
Comment (simonlbn) [inline examples/led_7seg_ht16k33/main.c]: Done.

Ah, that makes sense. What does the "O" in OSTR() stand for?

File: examples/led_7seg_ht16k33/main.c (hunk 9 above)
Comment (simonlbn) [inline examples/led_7seg_ht16k33/main.c]: Done.

File: examples/led_7seg_ht16k33/main.c (hunk 8 above)
Comment (simonlbn) [inline examples/led_7seg_ht16k33/main.c]: Ah, I thought I had fixed all of those ("old" emacs config used tab). Done.

File: src/drivers/led_7seg_ht16k33.c (hunk 10 above)
Comment (simonlbn) [inline src/drivers/led_7seg_ht16k33.c]: Ack, done.

File: src/drivers/led_7seg_ht16k33.h (hunk 11 above)
Comment (simonlbn) [inline src/drivers/led_7seg_ht16k33.h]: Done. Only in header files or also in the .c file (as I noticed one case where I hadn't done that in the .c file) ?

**Comment by simonlbn** [comment]
For naming, the ht16k33 chip is can be used with many LED configurations including matrixes (so numbers make less sense in that case). For example Adafruit has a LED matrix https://www.adafruit.com/products/3149. Personally I'm currently only really interested in displays which can show numbers etc. so I didn't want to make a totally generic driver right now as that would be much more complicated.

That said, I the ht16k33 is called LED Controller Driver so perhaps the "led" part of the driver name is a bit redundant and ht16k33_7seg would be shorter... another option entirely would be just to call the driver ht16k33, take a "display type" argument to the init function which takes for now only support something like ADAFRUIT_7SEG. I don't really have a strong opinion on which way would be better... do you have a preference?

PS. sorry about the delay in responding - not much spare time in the last week.

**Comment by eerimoq** [comment]
I don't have a strong opinion either. My knowledge of the display family is very limited and I don't have time to read up on it. I'll accept the current name, and possibly change it in the future.

**Pull request merged**